snap, hooks: make host-hunspell a system-files interface

bboozzoo opened the pull request:
The host-hunspell plug was a mount-control interface, which means that the snap could request snapd to set up a mount from/to a specific location. The problem is that the mount is visible in the host mount namespace, i.e. the mounts created a system wide. The mount-control interface was really intended to be used in Ubuntu Core oriented scenarios, when for instance a gadget mounts some external storage at a system wide location.

The particular case of the Firefox snap only really needs access to the hunspell dictionaries on the host. In this case a system-files interface allowing access to /usr/share/hunspell inside the host filesystem (exposed to the snap mount ns at /var/lib/snapd/hostfs) will be sufficient.

This will fix issues like LP#2059195

bboozzoo committed:
snap, hooks: make host-hunspell a system-files interface

The host-hunspell plug was a mount-control interface, which means that the snap
could request snapd to set up a mount from/to a specific location. The problem
is that the mount is visible in the host mount namespace, i.e. the mounts
created a system wide. The mount-control interface was really intended to be
used in Ubuntu Core oriented scenarios, when for instance a gadget mounts some
external storage at a system wide location.

The particular care of the Firefox snap only really needs access to the hunspell
dictionaries on the host. In this case a system-files interface allowing access
to /usr/share/hunspell inside the host filesystem (exposed to the snap mount ns
at /var/lib/snapd/hostfs) will be sufficient.

This will fix issues like LP#2059195

Signed-off-by: Maciej Borzecki <maciej.borzecki@canonical.com>

zyga reviewed (approved):
This looks good to me

alexmurray commented on snapcraft.yaml:
```diff
@@ -103,12 +103,9 @@ plugs:
-      where: $SNAP_COMMON/host-hunspell
-      persistent: true
-      options: [ro, bind, noatime, noexec]
+    interface: system-files
```
This will require an update to the firefox snap declaration in the store.

bboozzoo replied:
Yes. Do you know if the store can handle the case where the plug name is unchanged? Maybe it'd be better if I changed the name to say `host-usr-share-hunspell`?

bboozzoo replied:
I've renamed the plug to host-usr-share-hunspell to avoid potential name conflicts in snap declaration.

alexmurray replied:
Thanks, yep I think this is better.

bandali replied:
Submitted store request for review/approval:

[Request for auto-connection of firefox:host-usr-share-hunspell plug of system-files interface](https://forum.snapcraft.io/t/request-for-auto-connection-of-firefox-host-usr-share-hunspell-plug-of-system-files-interface/39619)

bandali replied:
FTR, the store request was granted, so this thread can be resolved.

bboozzoo committed:
snap: rename plug to host-usr-share-hunspell

Rename the plug to host-usr-share-hunspell to avoid potential issue with
conflicting names in the snap declaration.

Signed-off-by: Maciej Borzecki <maciej.borzecki@canonical.com>

seb128 commented:
Not a review but a side comment but we should start by targeting nightly and beta which are higher risk channels where we usually land changes to get testing before cherry-picking to stable

bandali commented:
Thanks for the PR.  From a quick look and test, it looks good.  I've pushed the change to the `nightly` and `beta` branches, and the corresponding builds are now pending manual review on the store.  Once the approval is granted, I'll test a few other scenarios as well, before merging this into `stable`.

ssbatema commented:
installs and runs on raspberry pi4b+ubuntu core22 (arm64), closing [52](https://github.com/canonical/firefox-snap/pull/52)!

Commit:
Restore disconnect-plug-host-hunspell hook and host-hunspell declaration

We need this to be able to unmount the existing mount point for the
last time, otherwise removal of the Firefox snap fails halfway through
with this error:

error: cannot perform the following tasks:
- Remove data for snap "firefox" (4111) (unlinkat /var/snap/firefox/common/host-hunspell/en_US.dic: read-only file system)

Commit:
Restore previous 'DICPATH' and add comments about hunspell setup

bandali commented:
Hi @bboozzoo, (all,)

As mentioned above, I'd cherry-picked your changes into `nightly` and `beta` earlier, and after closer examination/review and testing several scenarios since then, I found a few issues with the proposed changes:

- With the changes applied, attempting to remove the Firefox snap would fail midway through with the following error:
  ```
  error: cannot perform the following tasks:
  - Remove data for snap "firefox" (4111) (unlinkat /var/snap/firefox/common/host-hunspell/en_US.dic: read-only file system)
  ```
  Which appears to be due to the leftover mount point after the switch away from mount-control, and snapd not cleaning it up.
  As such, I added a commit to restore the mount-control declaration to `snapcraft.yaml` for the time being and also restored `snap/hooks/disconnect-plug-host-hunspell` to clean up after ourselves when disconnecting the plug.  With this, running `sudo snap remove firefox` seems to be all good again and the removal completes successfully.

- Your patch set `DICPATH` to `/var/lib/snapd/hostfs/usr/share/hunspell`, but it was previously deliberately set to not point directly to the host dictionaries, mainly to avoid potential incompatibility issues between differing versions of hunspell in the snap and outside the snap on the host.  So I added a commit restoring the previous `DICPATH`, and adapting `snap/hooks/post-refresh` accordingly along with comments explaining the setup and the reasons for it.

- Lastly, related to the first point, since your proposed changes had been live on `nightly` and `beta` for a few weeks, we can assume a nonzero number of people already their Firefox snap updated to those revisions, meaning their `host-hunspell` plug was disconnected without proper cleanup (since the initial proposed changes dropped `snap/hooks/disconnect-plug-host-hunspell`) so they would run into the aforementioned error if they were to attempt to remove the Firefox snap.  For that reason, I added a [third commit](https://github.com/canonical/firefox-snap/commit/b389ce47a561746e9593be6c72eaeed7c1b73990) to attempt a cleanup as part of `snap/hooks/post-refresh`.  I added the commit only to `nightly` and `beta`, and *not* this PR, since none of these changes are merged into `stable` yet, and so I believe for `stable` doing the cleanup only in `snap/hooks/disconnect-plug-host-hunspell` should be sufficient.

Thanks again for the patch, and thanks in advance for reviewing my proposed changes.

EDIT: also cc @lissyx

bboozzoo reviewed (commented):
LGTM, thanks for the update and extensive comments!

seb128 reviewed (approved):
Thanks, if it's fine by the snapd reviewers I'm also +1 :)

Commit:
Add needed host-usr-share-hunspell plug for the post-refresh hook

bandali commented:
Thanks for the reviews folks.

Per feedback from testing by Mozilla QA folks, it turned out that this was not working properly (even though I thought it did at some point, but perhaps I was misremembering), and with some digging the culprit turned out to be the post-refresh hook failing to set up the dictionary symlinks. With further debugging and looking in the snapd logs, I found out that the issue was due to the post-refresh hook not having sufficient permission to access the host dictionaries. Simply adding the `host-usr-share-hunspell` plug for the post-refresh hook resolved the issue (thanks @zyga!). I pushed the commit directly to `nightly` and `beta`, and added it to this PR as well.

I believe we're nearly ready to land this to stable as well. I'll request another round of testing by Mozilla QA, and if everything looks good I'll aim to land this in the coming weeks along with one of the new major releases.

bandali commented:
With Mozilla QA having done verification for this, I've now merged this into `stable` for the 129 major release RCs. (The change was already merged for `beta` and `nightly` Previously.)

Both myself and Mozilla QA will do further spotcheck tests this week once the 129 RC snap builds are available and next week when they're promoted to stable and officially released to monitor and double-check that things continue to work as expected.

Thanks all!
